python/lsst/daf/butler/datastores/chainedDatastore.py

```diff
@@ -305,97 +324,53 @@ def put(self, inMemoryDataset, ref):
-        at the end.  If a relative path is given then it is assumed the file
-        is already inside the child datastore.
+    def _prepIngest(self, *datasets: FileDataset, transfer: Optional[str] = None) -> _IngestPrepData:
+        # Docstring inherited from Datastore._prepIngest.
+        if transfer is None or transfer == "move":
+            raise NotImplementedError("ChainedDatastore does not support transfer=None or transfer='move'.")
 
-        A transfer mode of None implies that the file is already within
-        each of the (relevant) child datastores.
-
-        Parameters
-        ----------
-        path : `str`
-            File path.  Treated as relative to the repository root of each
-            child datastore if not absolute.
-        ref : `DatasetRef`
-            Reference to the associated Dataset.
-        formatter : `Formatter` (optional)
-            Formatter that should be used to retreive the Dataset.  If not
-            provided, the formatter will be constructed according to
-            Datastore configuration.
-        transfer : str (optional)
-            If not None, must be one of 'move', 'copy', 'hardlink', or
-            'symlink' indicating how to transfer the file.  The new
-            filename and location will be determined via template substitution,
-            as with ``put``.  If the file is outside the datastore root, it
-            must be transferred somehow.
-
-        Raises
-        ------
-        NotImplementedError
-            If all chained datastores have no ingest implemented or if
-            a transfer mode of `None` is specified.
-        DatasetTypeNotSupportedError
-            The associated `DatasetType` is not handled by this datastore.
-        """
-        log.debug("Ingesting %s (transfer=%s)", ref, transfer)
-
-        # Confirm that we can accept this dataset
-        if not self.constraints.isAcceptable(ref):
-            # Raise rather than use boolean return value.
-            raise DatasetTypeNotSupportedError(f"Dataset {ref} has been rejected by this datastore via"
-                                               " configuration.")
-
-        if transfer is None:
-            raise NotImplementedError("ChainedDatastore does not support transfer=None")
-
-        # A "move" is sometimes a "copy"
-        moveIsCopy = False
-        if transfer == "move" and os.path.isabs(path):
-            moveIsCopy = True
+        def isDatasetAcceptable(dataset, *, name, constraints):
+            if not constraints.isAcceptable(dataset.ref):
+                log.debug("Datastore %s skipping ingest via configuration for ref %s", name, dataset.ref)
+                return False
+            else:
+                return True
 
-        notImplementedCounter = 0
-        notAcceptedCounter = 0
+        # Filter down to just datasets the chained datastore's own
+        # configuration accepts.
+        okForParent: List[FileDataset] = [dataset for dataset in datasets
```

Comment:
Raising that exception (in ingest) is now done exclusively by `Datastore.ingest`, which will check that the set of datasets actually ingested matches the set of datasets that the caller asked to be ingested and raise if that isn't true.  That's because we need `_prepIngest` to still return the list of datasets it *can* ingest even when it can't ingest all of them, and it can't do that if it raises.

Reply:
I see that now. Thanks.